### PR TITLE
Split some essential MatrixTransport functions into independent utils

### LINKED ISSUE #3291: Extract server discovery and shared room discovery / creation into utility function
## Abstract

Sub-issue from #3243.
In the initial implementation the MS needs to be able to join a fully federated global room to receive balance proofs from Raiden clients.
Extract the server discovery / login mechanism and discovery room creation mechanism from the MatrixTransport into a utility function, this will provide a temporary solution to the MS until #3252 is implemented and the full MatrixTransport can be used by the MS.

## Tasks

- [ ] Extract server discovery, login
- [ ] Extract shared room creation 

## Time estimation

1 Week

### PR DESCRIPTION
**Pull request by andrevmatos**
- `join_global_room` (from `join_discovery_room`)
- `login_or_register` (with `Signer`)
- `validate_userid_signature`

This PR makes these standalone functions, which depends only on `GMatrixClient` (now also pulled in from `raiden-libs`) and utils (like Signer), instead of `RaidenService` and other core classes instances (which was the case for the `MatrixTransport`, to be decoupled on #3252).

This also adds unit tests for these functions, as part of #3124 
These functions should be enough to write a client capable of interacting with Raiden's MatrixTransport network, thus closes #3291

### CODE REVIEW

**Comment by Dominik1999** [comment]
Thank you Andre!

File: raiden/network/transport/matrix/__init__.py [hunk 1]
```diff
@@ -1,31 +1,25 @@
 import json
-import re
```
Comment (ulope) [inline raiden/network/transport/matrix/__init__.py]: Kind of a nitpick: Would you mind renaming this to `transport.py` (or similar) and just importing the relevant bits to `__init__`?

File: raiden/network/transport/matrix/__init__.py [hunk 2]
```diff
@@ -1370,7 +1215,8 @@ def _set_room_id_for_address(self, address: Address, room_id: Optional[_RoomID]
 
         with self._account_data_lock:
             # no need to deepcopy, we don't modify lists in-place
-            _address_to_room_ids: Dict[AddressHex, List[_RoomID]] = self._client.account_data.get(
+            # type: Dict[AddressHex, List[_RoomID]]
+            _address_to_room_ids = self._client.account_data.get(
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/network/transport/matrix/__init__.py]: It's in the commit description: mypy complained about the generic data coming from `GMatrixClient`: `account_data: Dict[str, Any]` incompatible with specified type, so I had to take out the type annotations here

File: raiden/network/transport/matrix/__init__.py [hunk 3]
```diff
@@ -1432,11 +1278,12 @@ def _get_room_ids_for_address(
 
             return room_ids
 
-    def _leave_unused_rooms(self, _address_to_room_ids: Dict[AddressHex, List[_RoomID]]):
+    def _leave_unused_rooms(self, _address_to_room_ids):
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/network/transport/matrix/__init__.py]: Same with from above comment

File: raiden/network/transport/matrix/__init__.py [hunk 4]
```diff
@@ -319,15 +316,16 @@ def _http_retry_delay() -> Iterable[float]:
...
-
         # partner need to be in this dict to be listened on
         self._address_to_userids: Dict[Address, Set[str]] = defaultdict(set)
         self._address_to_presence: Dict[Address, UserPresence] = dict()
         self._userid_to_presence: Dict[str, UserPresence] = dict()
         self._address_to_retrier: Dict[Address, _RetryQueue] = dict()
 
-        self._discovery_room_alias = None
+        self._global_rooms: Dict[str, Optional[Room]] = {
+            suffix: None
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/network/transport/matrix/__init__.py]: Yes, that can be done, just did it like that thinking when we have e.g. a `send_global(message: Message, global_room_suffix: str)` method

[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/network/transport/matrix/__init__.py]: Changed.

File: raiden/network/transport/matrix/utils.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+    Returns:
+        matrix's Room instance linked to client
+    """
+    assert urlparse(client.api.base_url).netloc, 'Invalid client\'s homeserver url'
+    servers = [
+        urlparse(s).netloc
+        for s in ([client.api.base_url] + list(servers))  # client's own server first
+        if urlparse(s).netloc
+    ]
+    servers = list(OrderedDict.fromkeys(servers))  # dedupe, keep order
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/network/transport/matrix/utils.py]: Changed.

File: raiden/network/transport/matrix/transport.py [hunk 6]
```diff
@@ -1370,7 +1205,8 @@ def _set_room_id_for_address(self, address: Address, room_id: Optional[_RoomID]
 
         with self._account_data_lock:
             # no need to deepcopy, we don't modify lists in-place
-            _address_to_room_ids: Dict[AddressHex, List[_RoomID]] = self._client.account_data.get(
+            # type: Dict[AddressHex, List[_RoomID]]
```
Comment (ulope) [inline raiden/network/transport/matrix/transport.py]: Still don't really like this. We should look if there is a way to 'explain' things to mypy. But we can do that later.

Comment (andrevmatos) [reply raiden/network/transport/matrix/transport.py]: @ulope Implemented in #3427 , as well as splitting and tests for the server selection mechanism.

**Pull request merged**